tests/test_other.py

```diff
@@ -9241,6 +9241,30 @@ def test_f_exception(self, compile_flags, link_flags, expect_caught):
     result = self.run_js('a.out.js', assert_returncode=0 if expect_caught else NON_ZERO)
     self.assertContainedIf('CAUGHT', result, expect_caught)
 
+  def test_exceptions_with_closure_and_without_catching(self):
+    # using invokes will require setThrew(), and closure will error if it is not
+    # defined. this test checks that we define it even without catching any
+    # exceptions (if we did catch exceptions, that would include library code
+    # that would use setThrew() anyhow)
+    create_file('src.cpp', r'''
+      #include <stdio.h>
+      #include <emscripten.h>
+
+      struct A {
+        ~A() {
+          puts("~A");
+        }
+      };
+
+      int main() {
+        // Construct an instance of a class with a destructor, which will cause the
+        // use of invokes to ensure its destructor runs.
```

Comment:
A try-finally (`catch (...)`) doesn't quite work, because adding a catch will include the library code for a catch. We need an invoke without a catch, is the hard part... another option might be longjmp, but that's no clearer I think.

Reply:
Its its that hard to repro.. how did it show up in the wild?

Reply:
How does the static constructor/destructor trigger cause this to occur?

Reply:
LLVM emits an invoke for it, but not a catch, because there is no exception to catch. (I'm not sure why `catch (..)` emits a catch - seems like it doesn't need to. Maybe in optimized builds it doesn't in some cases...)

Here is the LLVM IR with the constructor: you can see an invoke, with no cxxabi code to catch or throw anything. Yes, this is pretty rare and hard to trigger, hence we only got a report now I guess...

```llvm
; ModuleID = 'a.cpp'
source_filename = "a.cpp"
target datalayout = "e-m:e-p:32:32-i64:64-n32:64-S128"
target triple = "wasm32-unknown-emscripten"

%struct.A = type { i8 }

$_ZN1AD2Ev = comdat any

@_ZTIi = external constant i8*
@.str = private unnamed_addr constant [3 x i8] c"~A\00", align 1

; Function Attrs: noinline norecurse optnone mustprogress
define hidden i32 @main() #0 personality i8* bitcast (i32 (...)* @__gxx_personality_v0 to i8*) !dbg !14 {
entry:
  %retval = alloca i32, align 4
  %a = alloca %struct.A, align 1
  %exn.slot = alloca i8*, align 4
  %ehselector.slot = alloca i32, align 4
  store i32 0, i32* %retval, align 4
  call void @llvm.dbg.declare(metadata %struct.A* %a, metadata !18, metadata !DIExpression()), !dbg !25
  %exception = call i8* @__cxa_allocate_exception(i32 4) #4, !dbg !26
  %0 = bitcast i8* %exception to i32*, !dbg !26
  store i32 5, i32* %0, align 16, !dbg !26
  invoke void @__cxa_throw(i8* %exception, i8* bitcast (i8** @_ZTIi to i8*), i8* null) #5
          to label %unreachable unwind label %lpad, !dbg !26

lpad:                                             ; preds = %entry
  %1 = landingpad { i8*, i32 }
          cleanup, !dbg !27
  %2 = extractvalue { i8*, i32 } %1, 0, !dbg !27
  store i8* %2, i8** %exn.slot, align 4, !dbg !27
  %3 = extractvalue { i8*, i32 } %1, 1, !dbg !27
  store i32 %3, i32* %ehselector.slot, align 4, !dbg !27
  %call = call %struct.A* @_ZN1AD2Ev(%struct.A* nonnull dereferenceable(1) %a) #4, !dbg !27
  br label %eh.resume, !dbg !27

eh.resume:                                        ; preds = %lpad
  %exn = load i8*, i8** %exn.slot, align 4, !dbg !27
  %sel = load i32, i32* %ehselector.slot, align 4, !dbg !27
  %lpad.val = insertvalue { i8*, i32 } undef, i8* %exn, 0, !dbg !27
  %lpad.val1 = insertvalue { i8*, i32 } %lpad.val, i32 %sel, 1, !dbg !27
  resume { i8*, i32 } %lpad.val1, !dbg !27

unreachable:                                      ; preds = %entry
  unreachable
}

; Function Attrs: nofree nosync nounwind readnone speculatable willreturn
declare void @llvm.dbg.declare(metadata, metadata, metadata) #1

declare i8* @__cxa_allocate_exception(i32)

declare void @__cxa_throw(i8*, i8*, i8*)

declare i32 @__gxx_personality_v0(...)

; Function Attrs: noinline nounwind optnone
define linkonce_odr hidden %struct.A* @_ZN1AD2Ev(%struct.A* nonnull returned dereferenceable(1) %this) unnamed_addr #2 comdat personality i8* bitcast (i32 (...)* @__gxx_personality_v0 to i8*) !dbg !28 {
entry:
  %this.addr = alloca %struct.A*, align 4
  store %struct.A* %this, %struct.A** %this.addr, align 4
  call void @llvm.dbg.declare(metadata %struct.A** %this.addr, metadata !29, metadata !DIExpression()), !dbg !31
  %this1 = load %struct.A*, %struct.A** %this.addr, align 4
  %call = invoke i32 @puts(i8* getelementptr inbounds ([3 x i8], [3 x i8]* @.str, i32 0, i32 0))
          to label %invoke.cont unwind label %terminate.lpad, !dbg !32

invoke.cont:                                      ; preds = %entry
  ret %struct.A* %this1, !dbg !34

terminate.lpad:                                   ; preds = %entry
  %0 = landingpad { i8*, i32 }
          catch i8* null, !dbg !32
  %1 = extractvalue { i8*, i32 } %0, 0, !dbg !32
  call void @_ZSt9terminatev() #6, !dbg !32
  unreachable, !dbg !32
}

declare i32 @puts(i8*) #3

declare void @_ZSt9terminatev()

attributes #0 = { noinline norecurse optnone mustprogress "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="generic" }
attributes #1 = { nofree nosync nounwind readnone speculatable willreturn }
attributes #2 = { noinline nounwind optnone "frame-pointer"="none" "min-legal-vector-width"="0" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="generic" }
attributes #3 = { "frame-pointer"="none" "no-trapping-math"="true" "stack-protector-buffer-size"="8" "target-cpu"="generic" }
attributes #4 = { nounwind }
attributes #5 = { noreturn }
attributes #6 = { noreturn nounwind }

!llvm.dbg.cu = !{!0}
!llvm.module.flags = !{!10, !11, !12}
!llvm.ident = !{!13}

!0 = distinct !DICompileUnit(language: DW_LANG_C_plus_plus, file: !1, producer: "clang version 13.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project f13893f66a228400bf9bdf14be425e3dc6da0034)", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, enums: !2, imports: !3, splitDebugInlining: false, nameTableKind: None)
!1 = !DIFile(filename: "a.cpp", directory: "/home/azakai/Dev/emscripten")
!2 = !{}
!3 = !{!4}
!4 = !DIImportedEntity(tag: DW_TAG_imported_declaration, scope: !0, entity: !5, file: !9, line: 51)
!5 = !DIDerivedType(tag: DW_TAG_typedef, name: "nullptr_t", scope: !7, file: !6, line: 56, baseType: !8)
!6 = !DIFile(filename: "cache/sysroot/include/c++/v1/__nullptr", directory: "/home/azakai/Dev/emscripten")
!7 = !DINamespace(name: "std", scope: null)
!8 = !DIBasicType(tag: DW_TAG_unspecified_type, name: "decltype(nullptr)")
!9 = !DIFile(filename: "cache/sysroot/include/c++/v1/stddef.h", directory: "/home/azakai/Dev/emscripten")
!10 = !{i32 7, !"Dwarf Version", i32 4}
!11 = !{i32 2, !"Debug Info Version", i32 3}
!12 = !{i32 1, !"wchar_size", i32 4}
!13 = !{!"clang version 13.0.0 (/b/s/w/ir/cache/git/chromium.googlesource.com-external-github.com-llvm-llvm--project f13893f66a228400bf9bdf14be425e3dc6da0034)"}
!14 = distinct !DISubprogram(name: "main", scope: !1, file: !1, line: 10, type: !15, scopeLine: 10, flags: DIFlagPrototyped, spFlags: DISPFlagDefinition, unit: !0, retainedNodes: !2)
!15 = !DISubroutineType(types: !16)
!16 = !{!17}
!17 = !DIBasicType(name: "int", size: 32, encoding: DW_ATE_signed)
!18 = !DILocalVariable(name: "a", scope: !14, file: !1, line: 13, type: !19)
!19 = distinct !DICompositeType(tag: DW_TAG_structure_type, name: "A", file: !1, line: 4, size: 8, flags: DIFlagTypePassByReference | DIFlagNonTrivial, elements: !20, identifier: "_ZTS1A")
!20 = !{!21}
!21 = !DISubprogram(name: "~A", scope: !19, file: !1, line: 5, type: !22, scopeLine: 5, flags: DIFlagPrototyped, spFlags: 0)
!22 = !DISubroutineType(types: !23)
!23 = !{null, !24}
!24 = !DIDerivedType(tag: DW_TAG_pointer_type, baseType: !19, size: 32, flags: DIFlagArtificial | DIFlagObjectPointer)
!25 = !DILocation(line: 13, column: 11, scope: !14)
!26 = !DILocation(line: 14, column: 9, scope: !14)
!27 = !DILocation(line: 15, column: 7, scope: !14)
!28 = distinct !DISubprogram(name: "~A", linkageName: "_ZN1AD2Ev", scope: !19, file: !1, line: 5, type: !22, scopeLine: 5, flags: DIFlagPrototyped, spFlags: DISPFlagDefinition, unit: !0, declaration: !21, retainedNodes: !2)
!29 = !DILocalVariable(name: "this", arg: 1, scope: !28, type: !30, flags: DIFlagArtificial | DIFlagObjectPointer)
!30 = !DIDerivedType(tag: DW_TAG_pointer_type, baseType: !19, size: 32)
!31 = !DILocation(line: 0, scope: !28)
!32 = !DILocation(line: 6, column: 11, scope: !33)
!33 = distinct !DILexicalBlock(scope: !28, file: !1, line: 5, column: 14)
!34 = !DILocation(line: 7, column: 9, scope: !28)
```

Reply:
I guess I don't understand why having C++ object on the stack wouldn't also cause an `invoke` with a cleanup landing pad for calling the destructor.. even without any try/catch.       Does this need to be static object or can it be stack allocated?

Reply:
ignore me I see it is stack allocated.  lgtm